cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java

```diff
@@ -156,53 +156,48 @@ public ProgramController run(final Program program, ProgramOptions options) {
     final Service mapReduceRuntimeService = new MapReduceRuntimeService(cConf, hConf, mapReduce, spec, context,
                                                                         program.getJarLocation(), locationFactory,
                                                                         streamAdmin, txSystemClient);
-    final ProgramController controller = new MapReduceProgramController(mapReduceRuntimeService, context);
-    controller.addListener(new AbstractListener() {
+    mapReduceRuntimeService.addListener(new Service.Listener() {
       @Override
-      public void init(ProgramController.State state, @Nullable Throwable cause) {
+      public void starting() {
         store.setStart(program.getId(), runId.getId(), TimeUnit.SECONDS.convert(System.currentTimeMillis(),
```

Comment:
Also, you can add method to `RunIds` to simplify the conversion. E.g. `RunIds.getTime(RunId runId, TimeUnit unit)`.